File: client/cli/src/pie_client_cli/list.py
"""List command implementation for the Pie CLI.

This module implements the `pie-cli list` subcommand for querying
all running inferlet instances from a Pie engine.
"""

from pathlib import Path
from typing import Optional

from . import engine


def truncate_with_ellipsis(s: str, max_chars: int) -> str:
    """Truncate a string to max_chars, adding ellipsis if truncated."""
    if len(s) <= max_chars:
        return s
    if max_chars < 3:
        return s[:max_chars]
    return s[:max_chars - 3] + "..."


def format_arguments(args: list[str]) -> str:
    """Format arguments with proper quoting to preserve grouping."""
    formatted = []
    for arg in args:
        if not arg or any(c.isspace() for c in arg):
            # Quote if empty or contains whitespace
            escaped = arg.replace('"', '\\"')
            formatted.append(f'"{escaped}"')
        else:
            formatted.append(arg)
    return " ".join(formatted)


def handle_list_command(
    config: Optional[Path] = None,
    host: Optional[str] = None,
    port: Optional[int] = None,
    username: Optional[str] = None,
    private_key_path: Optional[Path] = None,
    full: bool = False,
    long: bool = False,
) -> None:
    """Handle the `pie-cli list` command.
    
    1. Creates a client configuration from config file and command-line arguments
    2. Connects to the Pie engine server
    3. Queries for all live instances
    4. Displays the list of running inferlet instances
    """
    client_config = engine.ClientConfig.create(
        config_path=config,
        host=host,
        port=port,
        username=username,
        private_key_path=private_key_path,
    )
    
    client = engine.connect_and_authenticate(client_config)
    
    try:
        instances = engine.list_instances(client)
        
        if not instances:
            print("✅ No running instances found.")
            return
        
        plural = "" if len(instances) == 1 else "s"
        print(f"✅ Found {len(instances)} running instance{plural}:")
        print()
        
        # Column widths
        id_width = 36 if full else (8 if long else 4)
        status_width = 8
        cmd_width = 24
        args_width = 32 if full else (60 if long else 64)
        
        # Print header
        print(
            f"{'ID':<{id_width}}  {'STATUS':<{status_width}}  "
            f"{'COMMAND':<{cmd_width}}  {'ARGUMENTS':<{args_width}}"
        )
        
        # Print separator
        print(
            f"{'-' * id_width}  {'-' * status_width}  "
            f"{'-' * cmd_width}  {'-' * args_width}"
        )
        
        # Print each instance
        for inst in instances:
            # Format UUID based on display mode
            if full:
                uuid_display = inst.id
            elif long:
                uuid_display = inst.id[:8]
            else:
                uuid_display = inst.id[:4]
            
            # Format other fields
            status_display = inst.status
            cmd_display = truncate_with_ellipsis(inst.cmd_name, cmd_width)
            args_str = format_arguments(inst.arguments)
            args_display = truncate_with_ellipsis(args_str, args_width)
            
            print(
                f"{uuid_display:<{id_width}}  {status_display:<{status_width}}  "
                f"{cmd_display:<{cmd_width}}  {args_display:<{args_width}}"
            )
    
    finally:
        engine.close_client(client)

File: client/cli-rust/src/list.rs
//! List command implementation for the Pie CLI.
//!
//! This module implements the `pie-cli list` subcommand for querying
//! all running inferlet instances from a Pie engine.

use crate::engine;
use anyhow::{Context, Result};
use clap::Args;
use std::path::PathBuf;

/// Arguments for the `pie-cli list` command.
#[derive(Args, Debug)]
pub struct ListArgs {
    /// Path to a custom TOML configuration file.
    #[arg(long)]
    pub config: Option<PathBuf>,
    /// The network host to connect to.
    #[arg(long)]
    pub host: Option<String>,
    /// The network port to connect to.
    #[arg(long)]
    pub port: Option<u16>,
    /// The username to use for authentication.
    #[arg(long)]
    pub username: Option<String>,
    /// Path to the private key file to use for authentication.
    #[arg(long)]
    pub private_key_path: Option<PathBuf>,
    /// Display the full UUID.
    #[arg(long)]
    pub full: bool,
    /// Display the first 8 characters of the UUID.
    #[arg(long)]
    pub long: bool,
}

/// Handles the `pie-cli list` command.
///
/// This function:
/// 1. Reads configuration from the specified config file or default config
/// 2. Creates a client configuration from config and command-line arguments
/// 3. Attempts to connect to the Pie engine server
/// 4. Queries for all live instances
/// 5. Displays the list of running inferlet instances
pub async fn handle_list_command(args: ListArgs) -> Result<()> {
    let client_config = engine::ClientConfig::new(
        args.config,
        args.host,
        args.port,
        args.username,
        args.private_key_path,
    )?;

    let client = engine::connect_and_authenticate(&client_config)
        .await
        .context("Failed to connect to Pie engine")?;

    let instances = client
        .list_instances()
        .await
        .context("Failed to list instances")?;

    if instances.is_empty() {
        println!("✅ No running instances found.");
    } else {
        println!(
            "✅ Found {} running instance{}:",
            instances.len(),
            if instances.len() == 1 { "" } else { "s" }
        );
        println!();

        // Column widths for the table
        // UUID is typically 36 characters (with hyphens)
        let id_width = if args.full {
            36
        } else if args.long {
            8
        } else {
            4
        };
        const STATUS_WIDTH: usize = 8;
        let args_width = if args.full {
            56
        } else if args.long {
            84
        } else {
            88
        };

        // Print table header
        println!(
            "{:<id_width$}  {:<status_width$}  {:<args_width$}",
            "ID",
            "STATUS",
            "ARGUMENTS",
            id_width = id_width,
            status_width = STATUS_WIDTH,
            args_width = args_width
        );

        // Print separator line
        println!(
            "{}  {}  {}",
            "-".repeat(id_width),
            "-".repeat(STATUS_WIDTH),
            "-".repeat(args_width)
        );

        // Print each instance
        for instance in instances.iter() {
            // Use full UUID, first 8 characters, or first 4 characters based on the flag
            let uuid_display = if args.full {
                instance.id.clone()
            } else if args.long {
                instance.id.chars().take(8).collect::<String>()
            } else {
                instance.id.chars().take(4).collect::<String>()
            };

            // Format status
            let status_display = format!("{:?}", instance.status);

            // Format arguments with proper quoting to show grouping
            let args_str = format_arguments(&instance.arguments);
            let args_display = truncate_with_ellipsis(args_str, args_width);

            // Print row with proper padding
            println!(
                "{:<id_width$}  {:<status_width$}  {:<args_width$}",
                uuid_display,
                status_display,
                args_display,
                id_width = id_width,
                status_width = STATUS_WIDTH,
                args_width = args_width
            );
        }
    }

    Ok(())
}

/// Formats arguments with proper quoting to preserve grouping information.
///
/// Arguments containing spaces or special characters are quoted to show
/// they are a single argument. For example: ["-p", "Hello World"] becomes
/// `-p "Hello World"` instead of `-p Hello World`.
fn format_arguments(args: &[String]) -> String {
    args.iter()
        .map(|arg| {
            // Quote if the argument contains spaces or is empty
            if arg.is_empty() || arg.contains(char::is_whitespace) {
                format!("\"{}\"", arg.replace('"', "\\\""))
            } else {
                arg.clone()
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

/// Truncates a string to a maximum length, appending "..." if truncated.
fn truncate_with_ellipsis(mut s: String, max_display_chars: usize) -> String {
    let mut chars_scanned = 0;
    let mut truncation_byte_pos = None;
    let chars_before_ellipsis = max_display_chars.saturating_sub(3);

    for (byte_pos, _) in s.char_indices() {
        if chars_scanned == chars_before_ellipsis {
            truncation_byte_pos = Some(byte_pos);
        }
        chars_scanned += 1;
        if chars_scanned > max_display_chars {
            break;
        }
    }

    // If string fits within the limit, no truncation needed
    if chars_scanned <= max_display_chars {
        return s;
    }

    // Edge case: ellipsis itself is too long to fit
    if max_display_chars < 3 {
        return s.chars().take(max_display_chars).collect();
    }

    // Truncate and append ellipsis
    if let Some(byte_pos) = truncation_byte_pos {
        s.truncate(byte_pos);
        s.push_str("...");
    }
    s
}
